llvm/lib/Target/AIE/AIEHazardRecognizer.cpp

```diff
@@ -49,51 +49,65 @@ void FuncUnitWrapper::setFormatInterface(const AIEBaseMCFormats *Formats) {
 
 bool FuncUnitWrapper::operator==(const FuncUnitWrapper &Other) const {
   return Required == Other.Required && Reserved == Other.Reserved &&
-         Slots == Other.Slots;
+         Slots == Other.Slots && MemoryBanks == Other.MemoryBanks;
 }
 
 void FuncUnitWrapper::dump() const {
   const char *const Digits = "0123456789";
   const char *const Spacer = "-|";
   const int Upper = std::numeric_limits<InstrStage::FuncUnits>::digits - 1;
-  dbgs() << "Req     : ";
-  for (int J = Upper; J >= 0; J--)
-    dbgs() << ((Required & (1ULL << J)) ? Digits[J % 10] : Spacer[J % 10 == 0]);
-  dbgs() << " Slots : ";
-  for (int J = 9; J >= 0; J--)
-    dbgs() << ((Slots & (1ULL << J)) ? Digits[J] : '-');
+
+  auto printFU = [&](const std::string &FUName, InstrStage::FuncUnits FU) {
+    dbgs() << FUName;
+    for (int J = Upper; J >= 0; J--)
+      dbgs() << ((Required & (1ULL << J)) ? Digits[J % 10]
+                                          : Spacer[J % 10 == 0]);
+  };
+  auto printResource = [&](const std::string &ResourceName, uint64_t Resource) {
+    dbgs() << ResourceName;
+    for (int J = 9; J >= 0; J--)
+      dbgs() << ((Resource & (1ULL << J)) ? Digits[J] : '-');
+  };
+
+  printFU("Req     : ", Required);
+  printResource(" Slots : ", Slots);
+  printResource(" Memorybanks : ", MemoryBanks);
   if (!Reserved)
     return;
-  dbgs() << "\n\t   Rsrv : ";
-  for (int J = Upper; J >= 0; J--)
-    dbgs() << ((Reserved & (1ULL << J)) ? Digits[J % 10] : Spacer[J % 10 == 0]);
+  printFU("\n\t   Rsrv : ", Reserved);
 }
 
 void FuncUnitWrapper::clearResources() {
   IssueCount = 0;
   Required = 0;
   Reserved = 0;
   Slots = 0;
+  MemoryBanks = 0;
 }
 bool FuncUnitWrapper::isEmpty() const {
-  return Required == 0 && Reserved == 0 && Slots == 0;
+  return Required == 0 && Reserved == 0 && Slots == 0 && MemoryBanks == 0;
 }
 
 void FuncUnitWrapper::blockResources() {
   Required = ~0;
   Reserved = ~0;
   Slots = ~0;
+  // Since the HW stalls in the event of memory bank conflicts, we don't need to
+  // block the resource. It is overly conservative if we block all memory banks.
+  MemoryBanks = 0;
```

Comment:
That is resetting the bank bits. If there were some already set, I think we should keep them.